tools/autograd/gen_python_functions.py

```diff
@@ -293,436 +158,546 @@ def get_py_torch_functions(declarations):
+    'TensorList': 'tensorlist_n<{}>',
+    'DimnameList': 'dimnamelist',
+    'IntArrayRef': 'intlist',
+}
+
+UNPACK_WITH_DEFAULT_METHODS = {
+    'const Type &': 'scalartypeWithDefault',
+    'const THPLayout &': 'layoutWithDefault',
+    'const Device &': 'deviceWithDefault',
+}
+
+def parsed_arg_expr(arg, arg_index):
+    # e.g. for arg name 'foo', arg type 'bool', arg_index = 2, returns '_r.toBool(2)'
+    typename = arg['type']
+
+    default_init = arg.get('python_default_init')
+    if default_init is not None:
+        # Note: only introduced by make_python_binding_args
+        default_init = arg['python_default_init']
+        if typename not in UNPACK_WITH_DEFAULT_METHODS:
+            raise RuntimeError(
+                'type \'{}\' is not supported in python_default_init'.
+                format(typename))
+        unpack_with_default = UNPACK_WITH_DEFAULT_METHODS[typename]
+        return '_r.{}({}, {})'.format(unpack_with_default, arg_index, default_init)
+
+    size = arg.get('size')
+    if size is not None:
+        if typename not in UNPACK_WITH_SIZE_METHODS:
+            raise RuntimeError(
+                'type \'{}\' with definite size ({}) is not supported'.
+                format(typename, size))
+        unpack_with_size = UNPACK_WITH_SIZE_METHODS[typename].format(size)
+        return '_r.{}({})'.format(unpack_with_size, arg_index)
+
+    unpack = UNPACK_METHODS.get(typename)
+    if unpack is None:
+        raise RuntimeError('type \'{}\' is not supported'.format(typename))
+
+    return '_r.{}({})'.format(unpack, arg_index)
+
+
+# TODO make this part of something more general, or get rid of it
+def unpack_optional_dimname_list_hack(name, expr):
+    # optional<ArrayRef<T>> are special. The PythonArgParser returns an
+    # optional<vector<T>>, which cannot be implicitly converted to
+    # optional<ArrayRef<T>>. One needs to unwrap the optional and rewrap.
+    result = """\
+        auto __{name} = {expr};
+        c10::optional<{typ}> {name} = __{name} ? c10::make_optional({typ}(__{name}.value())) : c10::nullopt;
+    """.format(name=name, expr=expr, typ='DimnameList')
+    return [line.strip() for line in result.split('\n')]
+
+
+def parse_arg(arg, arg_index, unpack_to_local=False):
+    # get parsed rhs
+    expr = parsed_arg_expr(arg, arg_index)
+
+    # maybe unpack to local
+    name = arg['name']
+    typename = arg['type']
+    if typename == 'c10::optional<DimnameList>':
+        inits = unpack_optional_dimname_list_hack(name, expr)
+        expr = name
+    elif unpack_to_local:
+        inits = ['auto {} = {};'.format(name, expr)]
+        expr = name
+    else:
+        inits = []
+
+    return expr, inits
+
+
+#
+# schema type to cpp type conversions
+# some of these are to prevent dangling refs to temps, others are more obscure
+# TODO don't know if these fold into more general conversions somehere, hope so
+#
+
+TEMP_SAFE_CPP_DECL_TYPE = {
+    'Tensor': 'const Tensor &',
```

Comment:
I didn't look into this much, but this looks a little strange -- why does "Tensor" need a decl type transformation to be temp safe?

Reply:
I think I took the name (the conversion was already there) from the comment elsewhere in the file, about the dangling refs problem (Search for "Not good" :) But it didn't appear to be doing anything, so removing!